src/irr/asset/CPLYMeshFileLoader.cpp

```diff
@@ -19,15 +19,81 @@ namespace irr
 namespace asset
 {
 
-// constructor
-CPLYMeshFileLoader::CPLYMeshFileLoader()
+_IRR_STATIC_INLINE_CONSTEXPR const char* VERT_SHADER_CACHE_KEY = "irr/builtin/shader/loaders/ply/vertex";
+_IRR_STATIC_INLINE_CONSTEXPR const char* FRAG_SHADER_CACHE_KEY = "irr/builtin/shader/loaders/ply/fragment";
+
+const char* VERT_SHADER = R"===(
+#version 430 core
+layout(location = 0) in vec4 vPos;
+layout(location = 1) in vec4 vCol;
+layout(location = 3) in vec3 vNormal; 
+
+#include <irr/builtin/glsl/vertex_utils/vertex_utils.glsl>
+
+layout (set = 1, binding = 0, row_major, std140) uniform UBO {
+    irr_glsl_SBasicViewParameters params;
+} CamData;
+
+layout (location = 0) out vec3 color;
+layout (location = 1) out vec3 viewPos;
+layout (location = 2) out vec3 normal;
+
+void main()
+{
+    gl_Position = irr_glsl_pseudoMul4x4with3x1(irr_builtin_glsl_workaround_AMD_broken_row_major_qualifier_mat4x4(CamData.params.MVP), vPos);
+	viewPos = irr_glsl_pseudoMul3x4with3x1(irr_builtin_glsl_workaround_AMD_broken_row_major_qualifier_mat4x3(CamData.params.MV), vPos);
+	mat3 normalMat = irr_glsl_SBasicViewParameters_GetNormalMat(CamData.params);
+    normal = normalMat*normalize(vNormal);
+    color = vCol.xyz;
+}
+)===";
+
+const char* FRAG_SHADER = R"===(
+#version 430 core
+
+layout (location = 0) in vec3 color;
+layout (location = 1) in vec3 viewPos;
+layout (location = 2) in vec3 normal;
+
+layout(location = 0) out vec4 pixelColor;
+
+void main()
 {
+	// blinn-phong maybe?
+
+    pixelColor = vec4(color,1.0);
 }
+)===";
 
-CPLYMeshFileLoader::~CPLYMeshFileLoader()
+CPLYMeshFileLoader::CPLYMeshFileLoader(IAssetManager* _am) : m_assetMgr{ _am }
 {
+	auto createAndInstertShaderIntoCache = [&](const auto cacheKey, const auto shaderCode, const auto specialization)
+	{
+
+		std::string shader_key = cacheKey;
+		auto shader_unspec = core::make_smart_refctd_ptr<ICPUShader>(shaderCode);
+
+		ICPUSpecializedShader::SInfo specinfo({}, nullptr, "main", specialization);
+		auto shader_spec = core::make_smart_refctd_ptr<ICPUSpecializedShader>(std::move(shader_unspec), ICPUSpecializedShader::SInfo(specinfo));
+
+		asset::SAssetBundle bundle({ shader_spec });
+		m_assetMgr->changeAssetKey(bundle, shader_key);
+		m_assetMgr->insertAssetIntoCache(bundle);
+	};
+
+	auto createAndInsertPipelineLayoutIntoCache = [&]()
+	{
+
+	};
+
+	createAndInstertShaderIntoCache(VERT_SHADER_CACHE_KEY, VERT_SHADER, ICPUSpecializedShader::ESS_VERTEX);
+	createAndInstertShaderIntoCache(FRAG_SHADER_CACHE_KEY, FRAG_SHADER, ICPUSpecializedShader::ESS_FRAGMENT);
+
+	createAndInsertPipelineLayoutIntoCache(); // TODO
```

Comment:
I think there's already a pipeline layout for this type of shader (push constant matrix and no descriptor sets)

Reply:
yeah, we have something like that https://github.com/Crisspl/IrrlichtBAW/blob/shaders/src/irr/asset/CShaderIntrospector.cpp#L328

Reply:
but no default in IAssetManager.cpp